common/persistence/versionHistory.go

```diff
@@ -21,174 +21,517 @@
+	if item.version < lastItem.version {
+		return &shared.BadRequestError{Message: fmt.Sprintf(
+			"cannot update version history with a lower version %v. Last version: %v",
+			item.version, lastItem.version,
+		)}
+	}
+
+	if item.eventID <= lastItem.eventID {
+		return &shared.BadRequestError{Message: fmt.Sprintf(
+			"cannot add version history with a lower event id %v. Last event id: %v",
+			item.eventID, lastItem.eventID,
+		)}
+	}
+
+	if item.version > lastItem.version {
 		// Add a new history
-		v.History = append(v.History, currentItem)
+		v.items = append(v.items, item.Duplicate())
 	} else {
-		// item.version == lastItem.version && item.EventID > lastItem.EventID
-		// Update event  id
-		lastItem.EventID = currentItem.EventID
+		// item.version == lastItem.version && item.eventID > lastItem.eventID
+		// Update event ID
+		lastItem.eventID = item.eventID
 	}
 	return nil
 }
 
-// FindLowestCommonVersionHistoryItem returns the lowest version history item with the same version
-func (v *VersionHistory) FindLowestCommonVersionHistoryItem(remote VersionHistory) (VersionHistoryItem, error) {
-	localIdx := len(v.History) - 1
-	remoteIdx := len(remote.History) - 1
+// FindLCAItem returns the lowest common ancestor version history item
+func (v *VersionHistory) FindLCAItem(
+	remote *VersionHistory,
+) (*VersionHistoryItem, error) {
+
+	localIdx := len(v.items) - 1
+	remoteIdx := len(remote.items) - 1
 
 	for localIdx >= 0 && remoteIdx >= 0 {
-		localVersionItem := v.History[localIdx]
-		remoteVersionItem := remote.History[remoteIdx]
-		if localVersionItem.Version == remoteVersionItem.Version {
-			if localVersionItem.EventID > remoteVersionItem.EventID {
-				return remoteVersionItem, nil
+		localVersionItem := v.items[localIdx]
+		remoteVersionItem := remote.items[remoteIdx]
+
+		if localVersionItem.version == remoteVersionItem.version {
+			if localVersionItem.eventID > remoteVersionItem.eventID {
+				return remoteVersionItem.Duplicate(), nil
 			}
-			return localVersionItem, nil
-		} else if localVersionItem.Version > remoteVersionItem.Version {
+			return localVersionItem.Duplicate(), nil
+		} else if localVersionItem.version > remoteVersionItem.version {
 			localIdx--
 		} else {
 			// localVersionItem.version < remoteVersionItem.version
 			remoteIdx--
 		}
 	}
-	return VersionHistoryItem{}, &shared.BadRequestError{
-		Message: fmt.Sprintf("version history is malformed. No joint point found."),
+
+	return nil, &shared.BadRequestError{
+		Message: "version history is malformed. No joint point found.",
 	}
 }
 
-// IsAppendable checks if a version history item is appendable
-func (v *VersionHistory) IsAppendable(item VersionHistoryItem) bool {
-	if len(v.History) == 0 {
-		panic("version history cannot be empty")
+// IsLCAAppendable checks if a LCA version history item is appendable
+func (v *VersionHistory) IsLCAAppendable(
+	item *VersionHistoryItem,
+) bool {
+
+	if len(v.items) == 0 {
+		panic("version history not initialized")
```

Comment:
Way too many panics. Maybe you should start checking errors now...

Reply:
errors should be something that can happen,
panics is for definite code bug,

this is my current assumption

Reply:
but good point, too many panics